### ISSUE TITLE
This isn't measuring criticality, this is (sort of?) measuring health

### ISSUE THREAD

**Issue by kurtseifried** [ossf/criticality_score#30]
This isn't measuring criticality, this is (sort of?) measuring project health.

A criticality score implies, well, how critical something is. Is this thing needed to help keep the internet and major services running (e.g. Bind, PowerDNS, Python, PHP) or is it maybe not as critical?

Things like grep, sed, awk, bash, less obvious, but very much in the critical path of a lot of services functioning. Various software libraries taken for granted (e.g. XML processing, libjpeg) and so on. 

I'm not saying what you're measuring is a bad idea, but it doesn't strike me as measuring the criticalness of the software at all, just a general sense of the health of the project, which is fine, but please label it as such then.

**Comment by kurtseifried** [comment]
Also it's not clear to me how:

> Create a list of critical projects that the open source community depends on.


How is this done? Surveys? Dependency crawling? Expert analysis?

I'm also not clear on how the "criticality" data will support: 

> Use this data to proactively improve the security posture of these critical projects.

will it be used to justify asking the volunteer community for more resources? Will there simply be an attempt to game the metrics for better (more small commits?) or worse (fewer commits?). 

Again, any effort to improve Open Source Security is fine by me, but it's not clear that this will, or even what the actual plan is. Or what the measurement is supposed to show.

**Comment by inferno-chromium** [comment]
I think there is some confusion here, we are not talking about critical infrastructure services, but projects important/in use by open source community [see goals - https://github.com/ossf/criticality_score#goals]. You can still do it by adding additional weight paramaters on how a project is used in your organization/infrastructure (e.g. --params you can pass additional data point on how critical it is your infrastructure). And this is not talking about health, it is talking about reliance of users on your OSS projects (and that is what we define criticality). E.g. metric - dependents_count tells how other users use your OSS package in their project; user opening new issue and frequent updates reflects dependency of users on project, etc.

Also see definition - "A project's criticality score defines the influence and importance of a project."

**Comment by kurtseifried** [comment]
I also wasn't talking about critical infrastructure services. I'm not sure why you also changed the title. 

Also as a concrete example of what dpoes "criticality" mean, how about timezone data:

https://tech.slashdot.org/story/11/03/03/2143249/timezone-maintainer-retiring
https://tools.ietf.org/html/draft-lear-iana-timezone-database-02.html

You don't appear to list that in the top 200 projects. Is that something that would be counted?

**Comment by dlorenc** [comment]
I don't think there's a single definition of criticality across everyone. I see reaching a shared understanding as part of the fun here. If I were to try to put it into words, I think I've got:

Criticality is a measure of how much trouble the world would be if a CVSS 10.0 vulnerability got discovered in a particular piece of software.

This obviously can't be measured directly, so all of these metrics and algorithms are a rough proxy on that.

**Comment by dlorenc** [comment]
> Also as a concrete example of what dpoes "criticality" mean, how about timezone data:

Yep! I'd definitely consider tzdata critical! Same for stuff like ca-certs.

**Comment by inferno-chromium** [comment]
> I also wasn't talking about critical infrastructure services. I'm not sure why you also changed the title.
> 
> Also as a concrete example of what dpoes "criticality" mean, how about timezone data:
> 
> https://tech.slashdot.org/story/11/03/03/2143249/timezone-maintainer-retiring
> https://tools.ietf.org/html/draft-lear-iana-timezone-database-02.html
> 
> You don't appear to list that in the top 200 projects. Is that something that would be counted?

Right now, we only look at projects hosted on github, but we do have plans to look beyond that. I am not refuting your point, there are a lot of critical open source we should be looking, so need your ideas to look to create a list in a more automated fashion. Ideas? Lets work together. If you can create this manually, we would happy to ingest that in some way as well.

**Comment by kurtseifried** [comment]
Can I suggest another question you are all forgetting to ask?

What happens if that piece of code disappears from its normal home (e.g. GitHub or also things like NPM/PEAR/CPAN/etc)?

As for providing data I could easily hand you a dependency node graph for say Fedora or Debian but that's not very helpful until you actually define what problem you're trying to solve.

And that's what I'm asking: what specific problem is this effort trying to solve? E.g. reading the "We need ways to connect critical projects we all rely on with organizations that can provide them with support." cool, is there a list of these orgs that can provide support and more importantly, are willing to actually allocate resources to do so? Do we have some list of people clamoring to help?

**Comment by inferno-chromium** [comment]
Feel free to change title back if this does not seem correct. Our focus is definitely not on project health, but identify projects important for community.

> Can I suggest another question you are all forgetting to ask?
> 
> What happens if that piece of code disappears from its normal home (e.g. GitHub or also things like NPM/PEAR/CPAN/etc)?

We dont have an answer here. Can you expand on what problem you want to solve here.

> 
> As for providing data I could easily hand you a dependency node graph for say Fedora or Debian but that's not very helpful until you actually define what problem you're trying to solve.

Lets track this one in https://github.com/ossf/criticality_score/issues/23

> 
> And that's what I'm asking: what specific problem is this effort trying to solve? E.g. reading the "We need ways to connect critical projects we all rely on with organizations that can provide them with support." cool, is there a list of these orgs that can provide support and more importantly, are willing to actually allocate resources to do so? Do we have some list of people clamoring to help?

We are working as part of OpenSSF foundation, which is a consortium of companies listed here - https://openssf.org/about/members/. We will be working together to see how we can help various open source projects and what kind of help they need. This can be discussed more in the regular "Securing critical open source projects" working group, see information here - https://github.com/ossf/wg-securing-critical-projects

**Comment by kurtseifried** [comment]
You are looking at "criticality" and making noises about "security". I assume part of this would then be the classical CIA triad, Confidentiality (not usually applicable, unless people want to do secret squirrel security embargoes), Integrity (hugely applicable, we don't want backdoors/etc) and Availability, is the source code available when we need it? Which is kind of a cornerstone to this whole thing.

**Comment by inferno-chromium** [comment]
> You are looking at "criticality" and making noises about "security". I assume part of this would then be the classical CIA triad, Confidentiality (not usually applicable, unless people want to do secret squirrel security embargoes), Integrity (hugely applicable, we don't want backdoors/etc) and Availability, is the source code available when we need it? Which is kind of a cornerstone to this whole thing.

This is just initial work, we want community like you to participate and help us here with ideas. Do you have any ideas to objectively calculate security impact given a repo url OR have data points that can we can add to score ? We definitely care about security and want to account for it somehow.
